### PR TITLE
ExtHost - Documents: Hook up initial document load

### PR DESCRIPTION
**Pull request by bryphe**
__Issue:__ When opening an existing file, the contents from the open wouldn't get synchronized to the extension host.

We would send an update, but we would send with an endline that didn't exist yet - which the extension wouldn't handle. For an initial update, we should send our lines but at a start / end one-based position of `(line 1, column 1)` so that the insertion of the initial update is handled correctly.

Along the way, I stumbled across a couple of issues with our buffer updates - we should push some fixes (and improve the granularity / typing of updates we get from `reason-libvim`), so I also added more test coverage in our integration tests around buffer synchronization. This is something that must _always_ work.

### CODE REVIEW

File: src/Core/Types.re [hunk 1]
```diff
@@ -102,6 +102,8 @@ module Position = {
 module BufferUpdate = {
```
Comment (bryphe) [inline src/Core/Types.re]: Move this out of Core/Types

File: src/Store/VimStoreConnector.re [hunk 2]
```diff
@@ -317,17 +319,51 @@ let start =
...
           (),
         );
 
-      dispatch(Model.Actions.BufferUpdate(bu));
+      let shouldApply =
+        bufferMaybe
+        |> Option.map(Model.Buffer.shouldApplyUpdate(bu))
+        |> Option.value(~default=true);
+
+      if (shouldApply) {
```
Comment (bryphe) [inline src/Store/VimStoreConnector.re]: Add comment that we should be handling the buffer updates better from `reason-libvim` - in other words, we shouldn't _need_ this logic at this layer

File: src/Core/BufferUpdate.re [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  let ret: t = {
+    isFull,
+    id,
+    startLine: Index.OneBasedIndex(startLine),
+    endLine: Index.OneBasedIndex(endLine),
+    lines,
+    version,
+  };
+  ret;
+};
```
Comment (glennsl) [inline src/Core/BufferUpdate.re]: Why not just use `Index.t` here?

Also, `{ let ret: {isFull, ...}; ret; }` is equivalent to `{isFull, ...}`

File: src/Store/VimStoreConnector.re [hunk 4]
```diff
@@ -317,17 +319,59 @@ let start =
...
       Log.info("Vim - Buffer update: " ++ string_of_int(update.id));
       open Core.Types;
+      open Model.State;
+
+      let isFull = update.endLine == (-1);
+
+      let bufferMaybe =
+        Some(getState())
+        |> Option.map(state => state.buffers)
+        |> Option.bind(Model.Buffers.getBuffer(update.id));
```
Comment (glennsl) [inline src/Store/VimStoreConnector.re]: Isn't this equivalent to

```reason
getState().buffers |> Model.Buffers.getBuffer(update.id);
```
?

File: src/Store/VimStoreConnector.re [hunk 5]
```diff
@@ -317,17 +319,59 @@ let start =
       open Vim.BufferUpdate;
       Log.info("Vim - Buffer update: " ++ string_of_int(update.id));
       open Core.Types;
+      open Model.State;
+
+      let isFull = update.endLine == (-1);
+
+      let bufferMaybe =
```
Comment (glennsl) [inline src/Store/VimStoreConnector.re]: The convention is usually `maybeBuffer` I think. At least from what I've seen. It also reads better to me.

Good descriptive naming either way though :+1:

File: src/Store/VimStoreConnector.re [hunk 6]
```diff
@@ -317,17 +319,59 @@ let start =
...
+      // In particular, with undo / redo, we get multiple updates - a full update (version+2) and
+      // an incremental update (version+1) _after_ the full update. If we apply the incremental update
+      // after the full update, ignoring the version, we'll get incorrect results.
+      //
+      // The fix really belongs in reason-libvim - we should always be trust the order we get from the updates,
+      // and any of this filtering or manipulation of updates should be handled and tested there.
+      let shouldApply =
+        bufferMaybe
+        |> Option.map(Model.Buffer.shouldApplyUpdate(bu))
+        |> Option.value(~default=true);
```
Comment (glennsl) [inline src/Store/VimStoreConnector.re]: This feels a bit convoluted. How about `Option.map(Model.Buffer.shouldApplyUpdate(bu), bufferMaybe) != Some(false)`. Isn't that at least a little bit clearer?

Also, just opening `Oni_Model` at the top would remove quite a bit of noise. opening modules that only contain module aliases is usually fine.

File: src/Store/VimStoreConnector.re (hunk 4 above)
Comment (bryphe) [reply src/Store/VimStoreConnector.re]: Sure is... don't know what I was thinking 😄

File: src/Store/VimStoreConnector.re (hunk 6 above)
Comment (bryphe) [reply src/Store/VimStoreConnector.re]: Sounds good - I think the confusing aspect of the logic is that you still have to think about the `None` case. But it is cleaner. And hopefully the confusing part we can go away by fixing the root-cause bug in reason-libvim.

File: src/Core/BufferUpdate.re (hunk 3 above)
Comment (bryphe) [reply src/Core/BufferUpdate.re]: Sounds good - I switched to just `create` (and cleaned up the unnecessary `ret`)

**Pull request merged**